rpi/docs/config.txt
The following changes have been made to config.txt - those marked 'required for "NextPi Compatability"' are only needed on actual "Deployed Devices" in real nexts.

gpu_mem_512=16
	Forces GPU to only take 16Meg of Memory. This may cause problems for people wanting to do high texture 3D work later, but we can cross that bridge when we come to it...

dtparam=i2s
	Enable the I2S ALT mode on the Broadcom Silicon - required for "NextPi Compatability"

enable_uart=1
	Enable the UART ALT mode on the Broadcome Silicon - required for "NextPi Compatability"

dtoverlay=specnext-dac
	Enables the I2S TOSLink-alike soundcard in the Pi, and ALSA - required for "NextPi Compatability"

dtoverlay=rpi3-bt-miniuart 
	Swaps the MiniUART and the Real (fast) UART on machines with Wifi/Bluetooth. Technically this is not supported on "real NextPis", but has been added as a convenience for people who cannot get Pi0, and only get Pi0W.  UTTERLY UNTESTED ON Pi0W.


docs/config/config.txt
The config.ini file
-------------------
Author:  Phoebus R. Dokos (based on documentation by: Garry Lancaster &
                           Allen Allbright)
Date:	 2019-06-24
License: CC BY-NC-SA
Version: 1.1

Changelog
---------
2018-09-24: 1.0 Initial version
2019-06-24: 1.1 Major update due to FW changes

Introduction
============
Configuration of your ZX Spectrum Next computer can be done in two ways:
1 a. Via the Boot Screen (by pressing <SPACEBAR> while booting and then entering 
     the Editor Module with "E")
  b. Automatically upon boot specifically for the screen detection/adjustment
2. Via the config.ini file

The second method provides access to some settings not found on the regular ZX 
Spectrum Next Editor Module, adding a machine personality, changing the machine
type for said personality, provide alternate ROM images and even more 
importantly setting your monitor if it doesn't work out of the box.
Config.ini is a regular text file, located in the c:/tbblue/ folder of your boot
SD Card and it's read upon first boot after an F1 reset or cold boot.

Config.ini groups
=================
There are two basic groups of settings in the config.ini each one with it's own
sections/parts:
1. The global settings; or settings that apply to ALL the machine personalities
-and-
2. The per-personality settings (which also double as singular menu entries for 
each personality you setup)

Global Settings
===============
Global settings are set one-per-line with each setting's name followed by an '='
sign follwed by the value of the setting

Note: Starting with FW version 1.10b, if you modify the config.ini and enter 
non-acceptable values for each setting, the FW upon boot will replace these
values with the smallest acceptable one.

These are:
1. scandoubler
This sets the scandoubler ON (1) or OFF (0). Scandoubler is needed for modern 
hires monitors but should be turned off for older monitors. If you're using a 
PS/2 keyboard this setting can also be changed on the fly by pressing F2 other-
wise on a Spectrum membrane keyboard you'll need to press M1 (on the side of the
machine) + 2

Default setting:
scandoubler=1


2.50_60hz 
This sets the main frequency of the display to 50Hz(0) or 60Hz(1). If you're 
using a VGA monitor or are located in a country that uses 60Hz mains, chances 
are your monitor will lock at 60Hz. (But that's not all you need... look further
down in the per-personality settings section). 
If you're using a PS/2 keyboard this setting can also be changed on the fly by 
pressing F3 (or M1+3)

Default setting:
50_60hz=1


3. timex 
This enables Timex modes (HiRes and Multicolour) if set to (1)

Default setting:
timex=1

4.psgmode
psgmode sets the on board PSG mode to AY (0), YM (1) or disabled (2)

Default setting:
psgmode=0

5.stereomode
stereomode changes the stereophonic (Left-Center-Right) ordering of the PSG 
channels from ABC (0) to ACB (1)

Default setting:
stereomode=1

6.intsnd

Intsnd enables (1) or disables (0) the internal beeper.

Default setting:
intsnd=1

7.turbosound
Turbosound disables(0) or enables(1) the additional 2 AYs to give you either an 
option of 3 or 9 channels of PSG audio

Default setting:
turbosound=1

8.covox
The Covox setting enables (1) or disables (0) the onboard SpecDRUM / Covox 
Soundrive compatible DAC hardware.

Default setting:
covox=1

9.divmmc
This enables (1) or disables (0) the onboard divMMC. This won't affect boot but
will make the divMMC hardware unavailable to the selected personality. If this 
is enabled, then *if* the setting is *NOT* overriden by the per-personality 
settings as outlined in the following section, then a ROM image file named 
esxmmc.rom is expected to be located in c:/tbblue/.

Default setting:
divmmc=0

10. mf
This enables (1) or disables (0) the onboard MF-compatible hardware. It, too can
be overridden by the per-personality setting as outlined in the following 
section. If it is however enabled and the ROM filename is NOT overriden, then it
can be one of the following according to selected machine type. If machine type 
is 0 (48K) then the rom filename expected is mf.rom, if machine type is 1 (128K)
then the rom filename expected is mf128.rom and if it's 2 (+2/+3) then the 
expected rom filename is mf3.rom. These rom image files must be user provided 
and then placed in the c:/tbblue/ folder.

Default setting:
mf=0

11. joystick1
joystick1 (and joystick2 that follows) sets the first port joystick type to one
of the following options: 
0 = Sinclair 2 (67890)
1 = Kempston 1 (port 0x1F)
2 = Cursor (56780)
3 = Sinclair 1 (12345)
4 = Kempston 2 (port 0x37)
5 = MD 1 (3 or 6 button joystick port 0x1F)
-or-
6 = MD 2 (3 or 6 button joystick port 0x37)


Default setting:
joystick1=2

12. joystick2
This setting is as the previous one and accepts the same options (some of which 
are mutually exclusive i.e. if you select a Kempston 1 for joystick1 you cannot 
select a Kempston 1 for the second port as well)

Default setting:
joystick2=0

13. ps2
This setting rewires internally the PS/2 port to have the mouse connector as 
first PS/2 device(1) rather than the default which is keyboard(0). This 
facilitates plugging in a PS/2 mouse without a Y-splitter cable if you already 
have a membrane keyboard attached or a regular cased Next.

Default setting:
ps2=0

14. scanlines
This setting produces a cosmetic scanline-like effect of variable intensity from
totally disabled (0), to 75% (1), 50% (2) and 25% (3).
Scanlines are also controlled on the fly with the F8 key on the PS/2 keyboard 
(or M1+8)

Default setting:
scanlines=0

15. dac
The dac setting only applies to the Vtrucco development board and can be safely 
disregarded as it doesn't apply to the ZX Spectrum Next.

Default setting:
dac=1

16. turbo
turbo, enables(1) or disables (0) access to speeds over the standard 3.58Mhz. If
enabled, speed can be also overriden on the fly with the F7 key on a PS/2 
keyboard (or M1+7)

Default setting:
turbo=1

17. timing
timing sets a global screenmode setting from 0..7 for all menu items that have a
screen mode set to 8 and it's set automatically from the new testcard video
setting module upon boot. If a menu item contains a video mode different from 
8 then this setting is overridden.

18. default
this setting bridges the global configuration section with the per-personality 
configuration section/menu and selects which of the machine personalities that 
follow will be the one that boots by default. Entries are 0-based so the default
(9) actually means the 10th entry in the list

Default setting:
default=9

Per-Personality Settings
========================
The per-personality settings group serves three functions:
1. Create a boot menu entry per-line
2. Describe each machine personality's type as well as the display that's 
   connected to it
3. Override -if required- some global settings specifically for that personality

You will need one line starting with "menu=" per-personality up to a maximum of 
15 lines as follows:

menu="Name of machine",Type,Display,Main_ROM_Image_File,
      divMMC_ROM_Image_File,MF_ROM_Image_File

"Name of machine" is a meaninful name describing what this machine-personality 
                  is. 
                  Note: Make sure that the name is 32 characters long or shorter
                  as specifically for the Boot menu/Editor modules only standard
                  Spectrum resolution and fonts are supported.

e.g. ZX Spectrum 128K UK

Type can be one of the following: 0 (48K Spectrum), 
                                  1 (128K/+2 -grey- Spectrum), 
                                  2 (+2A/+3/Next), 
                                  3 (Soviet clone-style timings 128K Spectrum)

The type affects both the available memory, paging mechanism as well as ULA type
and timings of the machine personality. This setting can also be further 
modified after boot by modifying the appropriate Next hardware registers

Display can be a number from 0 to 6 for VGA or 7 for HDMI. This takes as a 
base the 50_60hz setting in the Global settings group and modifies the sync 
frequency in steps of approximately 1Hz to facilitate screen "lock" according to
your monitor. The higher the number the more the machine will speed-up although 
timings of all peripherals stay constant relative to one another. Setting 7 
(HDMI) completely modifies the timings to make the display compatible with 
modern HDMI monitors as well as changing the aspect ratio and border size 
(Especially for the 60Hz HDMI setting 7). 
Many HDMI monitors will work with a setting of 0 to 6 however ALL HDMI monitors 
will work with a setting of 7.
A setting of 8 allows the video testing boot module to set the screen according
to the user's input during boot (See timing setting above)

Main_ROM_Image_File is a ROM image file that's appropriate for the machine type
you've selected in the previous setting e.g. 48.rom for a standard 48K ZX 
Spectrum and which MUST be located in c:/tbblue/.

divMMC_ROM_Image_File. If this is set, then the global divMMC setting is 
overriden and the filename defined is used in lieu of the default esxmmc.rom. 
This rom image file is also expected to be found in c:/tbblue/ and lastly,

MF_ROM_Image_File, if this is set then the Global configuration for MF hardware 
is overriden, the hardware is enabled and the rom image file used is the one you
specify. This file should also be located in c:/tbblue/


e.g. menu=ZX Spectrum 128K,1,7,128Kalt.rom 
will add a boot menu entry that's enabling the 128K toastrack type with a bug 
fixed rom named 128Kal.rom and HDMI display type. If divMMC is enabled in global
settings this will require the presence of an esxDOS rom and if MF is enabled 
then it will require the presence of MF128.ROM



CHANGELOG
P.D     2019-01-27 - Core 2.00.26
                     Display: New Tilemap mode (640 x 256 and 320 x 256)
                     Documentation: New NextReg and I/O ports documentation
                                    Tilemap documentation
                     Software: NextZXOS v.2.00B
                               New .boot command by David Saphier (config 
                                                                    switcher)
                               New ScrollNutter version
                               New Tilemap demos (TM.nex)
                               Updated Spectron 2048 
                               Updated Kev Brady's utilities Overscan and
                                       Kempston
                               Updated Dizzy VII (now also includes .DSK image)
P.D.    2019-01-04 - Core 2.00.24 
                     Sprites:   
                     Completely New Sprite Engine (128 total sprites,
                     100 sprites per line, 4bit colour sprites, sprite scaling
                     - four sizes: x1,x2,x4,x8 - Anchored sprites -allowing
                     relative movement of sprites together)
                     Video:
                     Two new Layer colour mix modes -SLU 110(U+L) clamped and
                     SLU 111 colour = (U+L)-5 - allowing for colour darkening and
                     brightening for backgrounds
                     Audio:
                     Each AY can now be made monophonic
                     Fixed AY modules so they're completely silent when volume
                     is 0
                     Z80N:
                     ED 28: BSLA DE,B (shift DE left by B places - uses bits 
                                       4..0 of B only)  
                     ED 29: BSRA DE,B (arithmetic shift right DE by B places 
                                       - uses bits 4..0 of B only)  
                     ED 2A: BSRL DE,B (logical shift right DE by B places 
                                       - uses bits 4..0 of B only)  
                     ED 2B: BSRF DE,B (shift right DE by B places, filling from 
                                       left with 1s - uses bits 4..0 of B only)  
                     ED 2C: BRLC DE,B (rotate DE left by B places  
                                       - uses bits 3..0 of B only)  
                     (to rotate right, use B=16-places)
                     ED 98 JP (C) -- PC[13:0] = IN (C) << 6  
                    
                            
                     
P.D.    2018-10-31 - Core 2.00.11 (Fixes Timex alignment bug, fully disables
                     divMMC, beta priority L2 colour functions, Fixed DMA)
                     FW 1.10c
                     New SPREDIT
                     New Arkos 9 ch player dot command
P.D.    2018-09-24 - Added pinout information text under c:/docs/extra-hw/
P.D.    2018-09-19 - Initial commit of changelog for v.1.00



src/c/DotCommands/config/128/options.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>
#include <errno.h>
#include <arch/zxn.h>
#include <intrinsic.h>

#include "options.h"


// options search and sort

int sort_cmp_option(struct opt *a, struct opt *b)
{
   return stricmp(a->name, b->name);
}

int sort_opt_search(unsigned char *name, struct opt *a)
{
   if (a->type == OPT_TYPE_EXACT)
      return stricmp(name, a->name);
   
   return strnicmp(name, a->name, strlen(a->name));
}

// option arguments

static unsigned char *option_next_arg(unsigned char *idx, unsigned int argc, char **argv)
{
   if (++*idx < (unsigned char)argc) return strrstrip(strstrip(argv[*idx]));
   return 0;
}

static unsigned char *option_equal_arg(unsigned char *p)
{
   return strrstrip(strstrip(p));
}

static unsigned char *endp;

static unsigned char option_unsigned_number(unsigned char *p, unsigned int *res)
{
   if (p && *p)
   {
      errno = 0;
      *res = _strtou_(p, &endp, 0);
      
      if (errno == 0) return 1;  // success
   }
   
   return 0;                     // fail
}

// option error

unsigned int option_error(unsigned char *p)
{
   printf("Bad Parameter (%s)\n", p);
   return OPT_ACTION_OK;
}

// options

unsigned int option_exec_48k_helper(unsigned char t, unsigned char ct)
{
   ZXN_WRITE_REG(REG_MACHINE_TYPE, (ZXN_READ_REG(REG_MACHINE_TYPE) & 0x0f) | t);
   ZXN_WRITE_REG(REG_PERIPHERAL_3, (ZXN_READ_REG(REG_PERIPHERAL_3) & 0xbb) | ct);
   
   return OPT_ACTION_OK;
}

unsigned int option_exec_48k(void)
{
   // t=48, c=on, tmx=off
   return option_exec_48k_helper(0x80, 0x00);
}

unsigned int option_exec_128k(void)
{
   // t=128, c=on, tmx=off
   return option_exec_48k_helper(0xa0, 0x00);
}

unsigned int option_exec_plus3(void)
{
   // t=zxn, c=on, tmx=off
   return option_exec_48k_helper(0xb0, 0x00);
}

unsigned int option_exec_pentagon(void)
{
   // t=pent, c=off, tmx=off
   return option_exec_48k_helper(0xc0, 0x40);
}

unsigned int option_exec_zxn(void)
{
   // t=zxn, c=off, tmx=on
   return option_exec_48k_helper(0xb0, 0x44);
}

unsigned int option_exec_lock_helper(unsigned char val)
{
   if ((ZXN_READ_REG(REG_MACHINE_TYPE) & 0x6) == 0)
      return option_error("48k [un]lock");

   ZXN_WRITE_REG(REG_PERIPHERAL_3, (ZXN_READ_REG(REG_PERIPHERAL_3) & 0x7f) | val);
   return OPT_ACTION_OK;
}

unsigned int option_exec_lock(void)
{
   return option_exec_lock_helper(0);
}

unsigned int option_exec_unlock(void)
{
   return option_exec_lock_helper(0x80);
}

unsigned int option_exec_nextreg_helper(unsigned char *p)
{
   static unsigned int reg, val, mask;

   // nextreg reg,val[,mask]
   
   if (option_unsigned_number(p, &reg) && (reg < 256))
   {
      for (p = endp; isspace(*p) || (*p == ','); ++p) ;
      
      if (option_unsigned_number(p, &val) && (val < 256))
      {
         for (p = endp; isspace(*p) || (*p == ','); ++p) ;
         
         if (option_unsigned_number(p, &mask) && (mask < 256))
         {
            p = endp;
            val |= ZXN_READ_REG(reg) & mask;
         }
         
         if (*p == 0)
         {
            ZXN_WRITE_REG(reg, val);
            return OPT_ACTION_OK;
         }
      }
   }
   
   return option_error("nextreg");
}

unsigned int option_exec_nextreg(unsigned char *i, unsigned int argc, char **argv)
{
   return option_exec_nextreg_helper(option_next_arg(i, argc, argv));
}

unsigned int option_exec_nextreg_eq(unsigned char *i, unsigned int argc, char **argv)
{
   return option_exec_nextreg_helper(strchr(argv[*i], '=') + 1);
}

struct timing_type
{
   unsigned char *name;
   unsigned char value;
};

struct timing_type timing_types[] = {
   { "48", 0x80},
   { "48k", 0x80},
   { "128", 0xa0},
   { "128k", 0xa0},
   { "2", 0xa0},
   { "p", 0xc0},
   { "pent", 0xc0},
   { "pentagon", 0xc0},
   { "n", 0xb0},
   { "next", 0xb0},
   { "zxn", 0xb0},
   { "2A", 0xb0},
   { "2B", 0xb0},
   { "3", 0xb0},
   { "3E", 0xb0}
};

unsigned int option_exec_timing_helper(unsigned char *p)
{
   if (p && *p)
   {
      if (*p == '+') ++p;

      for (unsigned char i = 0; i != sizeof(timing_types) / sizeof(*timing_types); ++i)
      {
         if (stricmp(timing_types[i].name, p) == 0)
         {
            ZXN_NEXTREGA(REG_MACHINE_TYPE, (ZXN_READ_REG(REG_MACHINE_TYPE) & 0x0f) | timing_types[i].value);
            return OPT_ACTION_OK;
         }
      }
   }

   return option_error("timing");
}

unsigned int option_exec_timing(unsigned char *i, unsigned int argc, char **argv)
{
   return option_exec_timing_helper(option_next_arg(i, argc, argv));
}

unsigned int option_exec_timing_eq(unsigned char *i, unsigned int argc, char **argv)
{
   return option_exec_timing_helper(strchr(argv[*i], '=') + 1);
}

struct joy_type
{
   unsigned char *name;
   unsigned char value;
};

struct joy_type joy_types[] = {
   { "12345", 0x03 },
   { "s1", 0x03 },
   { "sinc1", 0x03 },
   { "sinclair1", 0x03 },
   { "67890", 0x00 },
   { "s2", 0x00 },
   { "sinc2", 0x00 },
   { "sinclair2", 0x00 },
   { "k", 0x01 },
   { "kemp", 0x01 },
   { "kempston", 0x01 },
   { "k1", 0x01 },
   { "kemp1", 0x01 },
   { "kempston1", 0x01 },
   { "k2", 0x04 },
   { "kemp2", 0x04 },
   { "kempston2", 0x04 },
   { "5678", 0x02 },
   { "c", 0x02 },
   { "cur", 0x02 },
   { "cursor", 0x02 },
   { "md1", 0x05 },
   { "md2", 0x06 }
};

unsigned int option_exec_joy_helper(unsigned char *p, unsigned char joy)
{
   static unsigned char mask, val;

   if (p && *p)
   {
      for (unsigned char i = 0; i != sizeof(joy_types) / sizeof(*joy_types); ++i)
      {
         // The compiler has not been modified to understand the size of
         // the new z80n instructions.  This loop has a nextreg instruction
         // that pushes the loop size out of range of a JR, causing an
         // assemble issue.  So NOPs have been manually inserted to increase
         // the size of the loop to avoid the JR substitution.

         intrinsic_nop();
         intrinsic_nop();
         
         if (stricmp(joy_types[i].name, p) == 0)
         {
            if (joy == 0)
            {
               mask = 0x37;
               val = (joy_types[i].value << 1) & 0x08;
               val |= (joy_types[i].value << 6) & 0xc0;
            }
            else
            {
               mask = 0xcd;
               val = (joy_types[i].value >> 1) & 0x02;
               val |= (joy_types[i].value << 4) & 0x30;
            }

            ZXN_NEXTREGA(REG_PERIPHERAL_1, (ZXN_READ_REG(REG_PERIPHERAL_1) & mask) | val);
            return OPT_ACTION_OK;
         }
      }
   }
   
   return option_error("joy");
}

unsigned int option_exec_joy0(unsigned char *i, unsigned int argc, char **argv)
{
   return option_exec_joy_helper(option_next_arg(i, argc, argv), 0); 
}

unsigned int option_exec_joy0_eq(unsigned char *i, unsigned int argc, char **argv)
{
   return option_exec_joy_helper(strchr(argv[*i], '=') + 1, 0);
}

unsigned int option_exec_joy1(unsigned char *i, unsigned int argc, char **argv)
{
   return option_exec_joy_helper(option_next_arg(i, argc, argv), 1);
}

unsigned int option_exec_joy1_eq(unsigned char *i, unsigned int argc, char **argv)
{
   return option_exec_joy_helper(strchr(argv[*i], '=') + 1, 1);
}

extern unsigned char original_cpu_speed;

unsigned int option_exec_turbo_35(void)
{
   original_cpu_speed = RTM_3MHZ;
   return OPT_ACTION_OK;
}

unsigned int option_exec_turbo_7(void)
{
   original_cpu_speed = RTM_7MHZ;
   return OPT_ACTION_OK;
}

unsigned int option_exec_turbo_14(void)
{
   original_cpu_speed = RTM_14MHZ;
   return OPT_ACTION_OK;
}

unsigned int option_exec_50hz(void)
{
   ZXN_NEXTREGA(REG_PERIPHERAL_1, ZXN_READ_REG(REG_PERIPHERAL_1) & 0xfb);
   return OPT_ACTION_OK;
}

unsigned int option_exec_60hz(void)
{
   ZXN_NEXTREGA(REG_PERIPHERAL_1, ZXN_READ_REG(REG_PERIPHERAL_1) | 0x04);
   return OPT_ACTION_OK;
}

struct scanline_type
{
   unsigned char *name;
   unsigned char value;
};

struct scanline_type scanline_types[] = {
   { "0", 0 },
   { "off", 0 },
   { "1", 0x03 },
   { "25", 0x03 },
   { "min", 0x03 },
   { "on", 0x02 },
   { "2", 0x02 },
   { "50", 0x02 },
   { "med", 0x02 },
   { "mid", 0x02 },
   { "medium", 0x02 },
   { "3", 0x01 },
   { "75", 0x01 },
   { "max", 0x01 },
   { "maximum", 0x01 }
};

unsigned int option_exec_scanline_helper(unsigned char *p)
{
   if (p && *p)
   {
      for (unsigned char i = 0; i != sizeof(scanline_types) / sizeof(*scanline_types); ++i)
      {
         if (stricmp(p, scanline_types[i].name) == 0)
         {
            ZXN_NEXTREGA(REG_PERIPHERAL_4, (ZXN_READ_REG(REG_PERIPHERAL_4) & 0xfc) | scanline_types[i].value);
            return OPT_ACTION_OK;
         }
      }
   }
   
   return option_error("scanline");
}

unsigned int option_exec_scanline(unsigned char *i, unsigned int argc, char **argv)
{
   return option_exec_scanline_helper(option_next_arg(i, argc, argv));
}

unsigned int option_exec_scanline_eq(unsigned char *i, unsigned int argc, char **argv)
{
   return option_exec_scanline_helper(strchr(argv[*i], '=') + 1);
}

struct contention_type
{
   unsigned char *name;
   unsigned char value;
};

struct contention_type contention_types[] =
{
   { "0", 0x40 },
   { "off", 0x40 },
   { "1", 0x00 },
   { "on", 0x00 }
};

unsigned int option_exec_contention_helper(unsigned char *p)
{
   if (p && *p)
   {
      for (unsigned char i = 0; i != sizeof(contention_types) / sizeof(*contention_types); ++i)
      {
         if (stricmp(p, contention_types[i].name) == 0)
         {
            ZXN_NEXTREGA(REG_PERIPHERAL_3, (ZXN_READ_REG(REG_PERIPHERAL_3) & 0xbf) | contention_types[i].value);
            return OPT_ACTION_OK;
         }
      }
   }
   
   return option_error("contention");
}

unsigned int option_exec_contention(unsigned char *i, unsigned int argc, char **argv)
{
   return option_exec_contention_helper(option_next_arg(i, argc, argv));
}

unsigned int option_exec_contention_eq(unsigned char *i, unsigned int argc, char **argv)
{
   return option_exec_contention_helper(strchr(argv[*i], '=') + 1);
}

struct ay_type
{
   unsigned char *name;
   unsigned char nextreg;
   unsigned char mask;
   unsigned char value;
};

// must have longer strings ahead of substrings

struct ay_type ay_types[] = {
   { "ts", REG_PERIPHERAL_3, 0xfd, 0x02 },
   { "turbosound", REG_PERIPHERAL_3, 0xfd, 0x02 },
   { "2", REG_PERIPHERAL_3, 0xfd, 0x02 },
   { "3", REG_PERIPHERAL_3, 0xfd, 0x02 },
   { "legacy", REG_PERIPHERAL_3, 0xfd, 0x00 },
   { "single", REG_PERIPHERAL_3, 0xfd, 0x00 },
   { "one", REG_PERIPHERAL_3, 0xfd, 0x00 },
   { "1", REG_PERIPHERAL_3, 0xfd, 0x00 },
   { "ay", REG_PERIPHERAL_2, 0xfc, 0x01 },
   { "ym", REG_PERIPHERAL_2, 0xfc, 0x00 },
   { "on", REG_PERIPHERAL_2, 0xfc, 0x00 },
   { "off", REG_PERIPHERAL_2, 0xfc, 0x03 },
   { "quiet", REG_PERIPHERAL_2, 0xfc, 0x03 },
   { "silent", REG_PERIPHERAL_2, 0xfc, 0x03 },
   { "mono", REG_PERIPHERAL_4, 0x1f, 0xe0 },
   { "m", REG_PERIPHERAL_4, 0x1f, 0xe0 },
   { "stereo", REG_PERIPHERAL_4, 0x1f, 0x00 },
   { "s", REG_PERIPHERAL_4, 0x1f, 0x00 },
   { "abc", REG_PERIPHERAL_3, 0xdf, 0x00 },
   { "acb", REG_PERIPHERAL_3, 0xdf, 0x20 }
};

unsigned int option_exec_aymode_helper(unsigned char *p)
{
   static unsigned char match;
   
   if (p && *p)
   {
      while (p && *p)
      {
         match = 0xff;
      
         for (unsigned char i = 0; i != sizeof(ay_types) / sizeof(*ay_types); ++i)
         {
            if (strnicmp(p, ay_types[i].name, strlen(ay_types[i].name)) == 0)
            {
               p += strlen(ay_types[i].name);
               while (isspace(*p)) ++p;
            
               if ((*p == 0) || (*p == ','))
                  match = i;
            
               while (isspace(*p) || (*p == ',')) ++p;
               break;
            }
         }
      
         if (match == 0xff)
            return option_error("aymode");
      
         ZXN_WRITE_REG(ay_types[match].nextreg, (ZXN_READ_REG(ay_types[match].nextreg) & ay_types[match].mask) | ay_types[match].value);
      }
      
      return OPT_ACTION_OK;
   }
   
   return option_error("aymode");
}

unsigned int option_exec_aymode(unsigned char *i, unsigned int argc, char **argv)
{
   return option_exec_aymode_helper(option_next_arg(i, argc, argv));
}

unsigned int option_exec_aymode_eq(unsigned char *i, unsigned int argc, char **argv)
{
   return option_exec_aymode_helper(strchr(argv[*i], '=') + 1);
}

struct ays_type
{
   unsigned char *name;
   unsigned char value;
};

struct ays_type ays_types[] = {
   { "m", 0xff },
   { "mono", 0xff },
   { "s", 0x00 },
   { "stereo", 0x00 }
};

unsigned int option_exec_ay_helper(unsigned char *p, unsigned char ay)
{
   static unsigned char mask;
   
   if (p && *p)
   {
      for (unsigned char i = 0; i != sizeof(ays_types) / sizeof(*ays_types); ++i)
      {
         if (stricmp(p, ays_types[i].name) == 0)
         {
            switch (ay)
            {
               case 0:
                  mask = 0x20;
                  break;
               
               case 1:
                  mask = 0x40;
                  break;
               
               default:
                  mask = 0x80;
                  break;
            }
            
            ZXN_NEXTREGA(REG_PERIPHERAL_4, (ZXN_READ_REG(REG_PERIPHERAL_4) & ~mask) |  (ays_types[i].value & mask));
            return OPT_ACTION_OK;
         }
      }
   }
   
   return option_error("ay");
}

unsigned int option_exec_ay0(unsigned char *i, unsigned int argc, char **argv)
{
   return option_exec_ay_helper(option_next_arg(i, argc, argv), 0);
}

unsigned int option_exec_ay0_eq(unsigned char *i, unsigned int argc, char **argv)
{
   return option_exec_ay_helper(strchr(argv[*i], '=') + 1, 0);
}

unsigned int option_exec_ay1(unsigned char *i, unsigned int argc, char **argv)
{
   return option_exec_ay_helper(option_next_arg(i, argc, argv), 1);
}

unsigned int option_exec_ay1_eq(unsigned char *i, unsigned int argc, char **argv)
{
   return option_exec_ay_helper(strchr(argv[*i], '=') + 1, 1);
}

unsigned int option_exec_ay2(unsigned char *i, unsigned int argc, char **argv)
{
   return option_exec_ay_helper(option_next_arg(i, argc, argv), 2);
}

unsigned int option_exec_ay2_eq(unsigned char *i, unsigned int argc, char **argv)
{
   return option_exec_ay_helper(strchr(argv[*i], '=') + 1, 2);
}

struct speaker_type
{
   unsigned char *name;
   unsigned char value;
};

struct speaker_type speaker_types[] = {
   { "0", 0x00 },
   { "off", 0x00 },
   { "1", 0x10 },
   { "on", 0x10 }
};

unsigned int option_exec_speaker_helper(unsigned char *p)
{
   if (p && *p)
   {
      for (unsigned char i = 0; i != sizeof(speaker_types) / sizeof(*speaker_types); ++i)
      {
         if (stricmp(p, speaker_types[i].name) == 0)
         {
            ZXN_NEXTREGA(REG_PERIPHERAL_3, (ZXN_READ_REG(REG_PERIPHERAL_3) & 0xef) | speaker_types[i].value);
            return OPT_ACTION_OK;
         }
      }
   }
   
   return option_error("speaker");
}

unsigned int option_exec_speaker(unsigned char *i, unsigned int argc, char **argv)
{
   return option_exec_speaker_helper(option_next_arg(i, argc, argv));
}

unsigned int option_exec_speaker_eq(unsigned char *i, unsigned int argc, char **argv)
{
   return option_exec_speaker_helper(strchr(argv[*i], '=') + 1);
}

struct timex_type
{
   unsigned char *name;
   unsigned char value;
};

struct timex_type timex_types[] = {
   { "0", 0x00 },
   { "off", 0x00 },
   { "1", 0x10 },
   { "on", 0x10 }
};

unsigned int option_exec_timex_helper(unsigned char *p)
{
   if (p && *p)
   {
      for (unsigned char i = 0; i != sizeof(timex_types) / sizeof(*timex_types); ++i)
      {
         if (stricmp(p, timex_types[i].name) == 0)
         {
            ZXN_NEXTREGA(REG_PERIPHERAL_3, (ZXN_READ_REG(REG_PERIPHERAL_3) & 0xfb) | timex_types[i].value);
            return OPT_ACTION_OK;
         }
      }
   }
   
   return option_error("timex");
}

unsigned int option_exec_timex(unsigned char *i, unsigned int argc, char **argv)
{
   return option_exec_timex_helper(option_next_arg(i, argc, argv));
}

unsigned int option_exec_timex_eq(unsigned char *i, unsigned int argc, char **argv)
{
   return option_exec_timex_helper(strchr(argv[*i], '=') + 1);
}

struct dac_type
{
   unsigned char *name;
   unsigned char value;
};

struct dac_type dac_types[] = {
   { "0", 0x00 },
   { "off", 0x00 },
   { "dis", 0x00 },
   { "disable", 0x00 },
   { "disabled", 0x00 },
   { "1", 0x08 },
   { "on", 0x08 },
   { "en", 0x08 },
   { "ena", 0x08 },
   { "enable", 0x08 },
   { "enabled", 0x08 }
};

unsigned int option_exec_dac_helper(unsigned char *p)
{
   if (p && *p)
   {
      for (unsigned char i = 0; i != sizeof(dac_types) / sizeof(*dac_types); ++i)
      {
         if (stricmp(p, dac_types[i].name) == 0)
         {
            ZXN_NEXTREGA(REG_PERIPHERAL_3, (ZXN_READ_REG(REG_PERIPHERAL_3) & 0xf7) | dac_types[i].value);
            return OPT_ACTION_OK;
         }
      }
   }
   
   return option_error("dac");
}

unsigned int option_exec_dac(unsigned char *i, unsigned int argc, char **argv)
{
   return option_exec_dac_helper(option_next_arg(i, argc, argv));
}

unsigned int option_exec_dac_eq(unsigned char *i, unsigned int argc, char **argv)
{
   return option_exec_dac_helper(strchr(argv[*i], '=') + 1);
}

struct dma_type
{
   unsigned char *name;
   unsigned char value;
};

struct dma_type dma_types[] = {
   { "zxn", 0 },
   { "z80", 0x40 }
};

unsigned int option_exec_dma_helper(unsigned char *p)
{
   if (p && *p)
   {
      for (unsigned char i = 0; i != sizeof(dma_types) / sizeof(*dma_types); ++i)
      {
         if (stricmp(p, dma_types[i].name) == 0)
         {
            ZXN_NEXTREGA(REG_PERIPHERAL_2, (ZXN_READ_REG(REG_PERIPHERAL_2) & 0xbf) | dma_types[i].value);
            return OPT_ACTION_OK;
         }
      }
   }
   
   return option_error("dma");
}

unsigned int option_exec_dma(unsigned char *i, unsigned int argc, char **argv)
{
   return option_exec_dma_helper(option_next_arg(i, argc, argv));
}

unsigned int option_exec_dma_eq(unsigned char *i, unsigned int argc, char **argv)
{
   return option_exec_dma_helper(strchr(argv[*i], '=') + 1);
}

unsigned int option_exec_help(void)
{
   flags.help = 1;
   return OPT_ACTION_OK;
}
